pkgs/build-support/dotnet/add-nuget-deps/default.nix

```diff
@@ -12,24 +12,55 @@
 }:
 
 {
+  /**
+    A list of nuget packages.
+
+    Should be a list of derivations or arguments to be applied to `fetchNupkg`.
+
+    ```nix
+    [
+      # Arguments for use fetchNupkg
+      { pname = "FosterFramework"; version = "0.1.15-alpha"; hash = "sha256-lM6eYgOGjl1fx6WFD7rnRi/YAQieM0mx60h0p5dr+l8="; }
+
+      # Or a derivation
+      (fetchNupkg { pname = "Microsoft.AspNetCore.App.Runtime.linux-x64"; version = "8.0.1"; hash = "sha256-QbUQXjCzr8j8u/5X0af9jE++EugdoxMhT08F49MZX74="; })
+    ]
+    ```
+
+    Either a JSON or nix file with the same structure can also be used.
+  */
   nugetDeps,
   overrideFetchAttrs ? x: { },
 }:
 fnOrAttrs: finalAttrs:
 let
   attrs = if builtins.isFunction fnOrAttrs then fnOrAttrs finalAttrs else fnOrAttrs;
 
-  deps =
-    if (nugetDeps != null) then
-      if lib.isDerivation nugetDeps then
-        [ nugetDeps ]
-      else if lib.isList nugetDeps then
-        nugetDeps
-      else
-        assert (lib.isPath nugetDeps);
-        callPackage nugetDeps { fetchNuGet = fetchNupkg; }
+  callNupkg = pkg: if lib.isDerivation pkg then pkg else fetchNupkg pkg;
+
+  loadDeps =
+    x:
+    if x == null then
+      [ ]
+    else if lib.isDerivation x then
+      [ x ]
+    else if builtins.isList x then
+      builtins.map callNupkg x
     else
-      [ ];
+      loadDeps (
+        if builtins.isFunction x then
+          lib.warn
+            "Nix-based NuGet lockfiles have been deprecated, migrate to JSON lockfiles. Deleting the existing lockfile and re-running fetch-deps will generate a JSON lockfile."
+            callPackage
+            x
+            { fetchNuGet = fetchNupkg; }
```

Comment:
In this case, since `lib.warn ""` is effectively an identity function, yes it does do the right thing.

_Technically_, the warning is printed when evaluating `callPackage` itself, instead of when evaluating the usage of `callPackage`. So it does effectively the same thing, it just evaluates in a different order.

<details><summary>Repl example</summary>
<p>

```sh
nix-repl> fn = x: "hi ${x}"

nix-repl> fn "there"
"hi there"

nix-repl> builtins.trace "warn!" (fn "there")
trace: warn!
"hi there"

nix-repl> (builtins.trace "warn!" fn) "there"
trace: warn!
"hi there"

nix-repl> builtins.trace "warn!" fn "there"
trace: warn!
"hi there"
```

</p>
</details>

Reply:
Fixed this to do the calls in the right order.